### ISSUE TITLE
incorrect compiler error "expression is not an integral constant expression" in enum declaration

### ISSUE THREAD

**Issue by djnz00** [llvm/llvm-project#113092]
clang 18 incorrectly reports an error declaring an enum as the result of a compile-time evaluated integer expression.

reduced test case below. clang version 18.1.8 Target: x86_64-pc-linux-gnu

```
#include <iostream>
#include <memory>

struct B { int i; };

struct A {
  constexpr A(int i) { std::construct_at(__builtin_addressof(b), i); }
  constexpr ~A() { std::destroy_at(__builtin_addressof(b)); }
  union { B b; };
};

consteval int i() { return A{42}.b.i; } // wrapper to circumvent compiler bug

int main() {
  // enum { I = i() };		// no error with both clang and gcc
  enum { I = A{42}.b.i };	// error with clang, but not with gcc
  std::cout << I << '\n';
}
```

**Comment by tbaederr** [comment]
The evaluation works just fine and returns the expected `42`, but we reject it anyway because we record that is has side effects:
https://github.com/llvm/llvm-project/blob/d1401822e2d2753bed3ac597a42cc0b261de40a4/clang/lib/Sema/SemaExpr.cpp#L17184

**Comment by tbaederr** [comment]
... and the side effect is coming from the `Cleanup` of `A` being non-trivial, i.e. `QualType::isDestructedType()` returns `true`.
